File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {

    public void backTracking(int openCount, int closeCount, StringBuilder stack, List<String> result, int n){
        if(openCount==closeCount && openCount==n){
            result.add(stack.toString());
            return;
        }

        if(openCount<n){
            stack.append('(');
            backTracking(openCount+1, closeCount, stack, result, n);
            stack.deleteCharAt(stack.length()-1);
        }

        if(closeCount<openCount){
            stack.append(')');
            backTracking(openCount, closeCount+1, stack, result, n);
            stack.deleteCharAt(stack.length()-1);
        }
    }

    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        backTracking(0, 0, new StringBuilder(), result, n);
        return result;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        stack = []
        def backtrack(openN, closedN):
            #base case - append to result if opencount == closedcount == n
            if openN == closedN == n:
                res.append("".join(stack.copy()))
                return
            #create combinations where open braces can be added -> if and only if openCount is less than n
            if openN < n:
                stack.append("(")
                backtrack(openN + 1, closedN)
                stack.pop()
            #create combinations where closed braces can be added -> if and only if closedCount is less than openCount 
            if closedN < openN:
                stack.append(")")
                backtrack(openN, closedN + 1)
                stack.pop()
        backtrack(0, 0)
        return res